packages/odyssey-storybook/src/components/odyssey-labs/DataComponents/dataFunctions.ts

```diff
@@ -26,98 +19,100 @@ export const filterData = ({
-          return value.some((arrayValue) => {
-            const filterValue =
-              typeof arrayValue === "object" ? arrayValue.value : arrayValue;
-            return row[id as keyof Person]
-              ?.toString()
-              .toLowerCase()
-              .includes(filterValue.toString().toLowerCase());
-          });
-        }
-
-        // In the custom filter examples, we provide a "starting letter"
-        // control that allows the user to filter by whether the
-        // first letter is a vowel or consonant
-        if (id === "startLetter" && typeof row.name === "string") {
-          const firstLetter = row.name[0]?.toLowerCase();
-          if (value === "vowel") return "aeiou".includes(firstLetter);
-          if (value === "consonant") return !"aeiou".includes(firstLetter);
-          return true;
-        }
-
-        // General filtering for other columns
-        return row[id as keyof Person]
-          ?.toString()
-          .toLowerCase()
-          .includes(value.toString().toLowerCase());
-      });
-    });
-  }
-
-  // Implement sorting
-  if (sort && sort.length > 0) {
-    filteredData.sort((a, b) => {
-      for (const { id, desc } of sort) {
-        const aValue = a[id as keyof Person];
-        const bValue = b[id as keyof Person];
-
-        if (aValue < bValue) return desc ? 1 : -1;
-        if (aValue > bValue) return desc ? -1 : 1;
-      }
+  const searchFiltered = search
+    ? data.filter((row) =>
+        Object.values(row).some((value) =>
+          value.toString().toLowerCase().includes(search.toLowerCase()),
+        ),
+      )
+    : data;
+
+  const columnFiltered = filters
+    ? searchFiltered.filter((row) =>
+        filters.every(({ id, value }) => {
+          if (value === null || value === undefined) {
+            return true;
+          }
+
+          if (Array.isArray(value)) {
+            return value.some((arrayValue) => {
+              const filterValue =
+                typeof arrayValue === "object" ? arrayValue.value : arrayValue;
+              return row[id as keyof Person]
+                ?.toString()
+                .toLowerCase()
+                .includes(filterValue.toString().toLowerCase());
+            });
+          }
+
+          if (id === "startLetter" && typeof row.name === "string") {
+            const firstLetter = row.name[0]?.toLowerCase();
+            if (value === "vowel") return "aeiou".includes(firstLetter);
+            if (value === "consonant") return !"aeiou".includes(firstLetter);
+            return true;
+          }
+
+          return row[id as keyof Person]
+            ?.toString()
+            .toLowerCase()
+            .includes(value.toString().toLowerCase());
+        }),
+      )
+    : searchFiltered;
+
+  const sorted =
+    sort && sort.length > 0
+      ? [...columnFiltered].sort((a, b) =>
+          sort.reduce((result, { id, desc }) => {
+            if (result !== 0) return result;
```

Comment:
```suggestion
            if (result !== 0) {
              return result;
            }
```

I don't want `goto fail` bugs. Also, it's easier to skim it this way.